cns/deviceplugin/server.go

```diff
@@ -0,0 +1,153 @@
+	address              string
+	logger               *zap.Logger
+	deviceCounter        deviceCounter
+	listAndWatchDoneChan chan struct{}
+	deviceCheckInterval  time.Duration
+}
+
+func NewServer(address string, logger *zap.Logger, deviceCounter deviceCounter, deviceCheckInterval time.Duration) *Server {
+	return &Server{
+		address:              address,
+		logger:               logger,
+		deviceCounter:        deviceCounter,
+		listAndWatchDoneChan: make(chan struct{}),
+		deviceCheckInterval:  deviceCheckInterval,
+	}
+}
+
+// Run starts the grpc server and blocks until an error or context is cancelled. Wait on Ready to know when the server is ready.
+func (s *Server) Run(ctx context.Context) error {
+	grpcServer := grpc.NewServer()
+	v1beta1.RegisterDevicePluginServer(grpcServer, s)
+	defer close(s.listAndWatchDoneChan)
+
+	l, err := net.Listen("unix", s.address)
+	if err != nil {
+		return errors.Wrap(err, "error listening on socket")
+	}
+	errChan := make(chan error, 2) //nolint:gomnd // disabled in favor of readability
+	go func() {
+		errChan <- grpcServer.Serve(l)
+	}()
+	defer grpcServer.Stop()
+
+	select {
+	case err := <-errChan:
+		return errors.Wrap(err, "error running grpc server")
+	case <-ctx.Done():
+		return nil
+	}
+}
+
+// Ready blocks until the server is ready
+func (s *Server) Ready(ctx context.Context) error {
+	c, err := grpc.DialContext(ctx, s.address, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock(), //nolint:staticcheck // TODO: Move to grpc.NewClient method
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			conn, err := (&net.Dialer{}).DialContext(ctx, "unix", addr)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to dial context")
+			}
+			return conn, nil
+		}),
+	)
+	if err != nil {
+		return errors.Wrap(err, "error dialing local grpc server")
+	}
+	if err := c.Close(); err != nil {
+		return errors.Wrap(err, "error closing connection to local grpc server")
+	}
+	return nil
+}
+
+func (s *Server) Allocate(_ context.Context, req *v1beta1.AllocateRequest) (*v1beta1.AllocateResponse, error) {
+	s.logger.Info("allocate request", zap.Any("req", *req))
+	resps := make([]*v1beta1.ContainerAllocateResponse, len(req.ContainerRequests))
+	for i, containerReq := range req.ContainerRequests {
+		resp := &v1beta1.ContainerAllocateResponse{
+			Envs: make(map[string]string),
+		}
+		for j := range containerReq.DevicesIDs {
+			resp.Envs[fmt.Sprintf("NIC_%d", j)] = containerReq.DevicesIDs[j]
+		}
+		resps[i] = resp
+	}
+	r := &v1beta1.AllocateResponse{
+		ContainerResponses: resps,
+	}
+	return r, nil
+}
+
+func (s *Server) ListAndWatch(_ *v1beta1.Empty, stream v1beta1.DevicePlugin_ListAndWatchServer) error {
+	// send the initial count right away
+	advertisedCount := s.deviceCounter.getDeviceCount()
+	devices := make([]*v1beta1.Device, advertisedCount)
+	for i := range devices {
+		devices[i] = &v1beta1.Device{
+			ID:     fmt.Sprintf("NIC-%d", i),
```

Comment:
agreed for consistency and best practices but these actually don't matter and are totally made up, they are effectively meaningless but the device plugin contract requires that we return _something_

Reply:
but as soon as they exist somebody will try to use them for something 😢

Reply:
Is there any way we can be actively hostile to someone trying to use them for something? Can we make them UUIDs?

Reply:
I think it's most likely to be us, which is why I suggest generating them consistently 😅

Reply:
I think there is some confusion here. Adding to what Matt was trying to explain earlier.
1) The device Ids are not stored locally in CNS. But rather they are managed / stored by K8s. Also they will not be used by anyone other than kubelet/k8s. 
2) We just need to return some unique device-Id for each nic resource. So here we are returning NIC-%d. Given the fact that we can only have a few nics per node, I don't think we really need to use UUID here. @timraymond 

Also, if we use UUIDs then, when/if the device count changes, we are sending the updated deviceIDs to the kubelet. Since we are not storing the device state / id locally, if we use UUID, we can't regenerate the deviceID / UUID. I think using the device count as deviceId suffix also helps us to generate them deterministically without storing them locally.  

3) Agreed that we should make them consistent, i.e. NIC-%d at all places. I can move the prefix to a constant.